js/app/widgets/webShareDialog.js

```diff
@@ -0,0 +1,263 @@
+const WebKit2 = imports.gi.WebKit2;
+
+const Lang = imports.lang;
+
+/* Template needs WebView type */
+GObject.type_ensure( WebKit2.WebView.$gtype);
+
+/**
+ * Class: WebShareDialog
+ *
+ * A dialog containing a webview with sessions cookies obtained from GOA
+ */
+var WebShareDialog = new Lang.Class({
+    Name: 'WebShareDialog',
+    Extends: Gtk.Dialog,
+
+    Properties: {
+        'provider': GObject.ParamSpec.string('provider',
+            'Provider',
+            'GOA provider type to use',
+            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT,
+            null),
+
+        'uri': GObject.ParamSpec.string('uri',
+            'URI to load',
+            'The URI to load in the webview',
+            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT,
+            null),
+
+        'redirect-uri': GObject.ParamSpec.string('redirect-uri',
+            'Redirect URI',
+            'The redirect URI to know when the operation ended',
+            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT,
+            null),
+
+        'mobile': GObject.ParamSpec.boolean('mobile',
+            'Mobile user agent',
+            'Use a mobile user agent for the webview',
+            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT_ONLY,
+            false),
+    },
+
+    Template: 'resource:///com/endlessm/knowledge/data/widgets/websharedialog.ui',
+    InternalChildren: [ 'stack', 'overlay', 'webview', 'spinner', 'accountbox' ],
+
+    _init: function (props) {
+        this._provider = null;
+        this._uri = null;
+
+        /* Create an unique cookie path for this instance */
+        this._cookies_path_init();
+
+        /* Get GOA client */
+        this._goa_client = Goa.Client.new_sync(null);
+
+        /* Chain Up */
+        this.parent(props);
+
+        if (this.mobile) {
+            let settings = this._webview.get_settings();
+            settings.set_user_agent("Mozilla/5.0 (EndlessOS not Android;) AppleWebKit/533.1 (KHTML, like Gecko) Version/4.0 Mobile");
+
+            /* Also use a mobile size */
+            this.default_width = 320;
+            this.default_height = 480;
+        }
+        else {
+            /* Based on facebook share dialog size */
+            this.default_width = 560;
+            this.default_height = 600;
+        }
+
+        /* Bind WebView and Dialog title */
+        this._webview.connect("notify::title", () => { this.title = this._webview.title; });
+
+        /* Handle spinner */
+        this._webview.connect('load-changed', this._onLoadChanged.bind(this));
+
+        /* Handle redirect-uri transaction end */
+        this._webview.connect('decide-policy', this._onDecidePolicy.bind(this));
+
+        /* Get GOA account for this.provider */
+        this._update_goa_account();
+
+        /* Set webview uri */
+        this._update_uri();
+    },
+
+    vfunc_destroy: function () {
+        this._cookies_path_remove();
```

Comment:
Just noticed this in passing — currently you can't do `vfunc_destroy` in GJS because it's impossible to execute the JS code here during garbage collection.

Reply:
ok, in any case case I changed this in favor of a regular signal connection on the webview destroy

Reply:
I'm not sure that will work either ... it depends on whether the destroy signal is emitted right away or deferred until the next iteration of the main loop.

Reply:
`GtkWidget::destroy` emission is synchronous inside the `GObjectClass.dispose()` implementation for GtkWidget.

I'm worried that you cannot connect to the destroy signal in GJS, though; that would break a lot of assumptions, and lead to massive leaks.